### ISSUE TITLE
Removing AllDifferentiableVariables prevents using differential learning rates

### ISSUE THREAD

**Issue by sgugger** [tensorflow/swift-apis#456]
It's very common to use different values of hyper-parameters for different layers in a model. For instance:
- using differentiable learning rates when fine-tuning a pretrained model (aka training deep layers with a lower learning rate than the last ones)
- using a weight decay of 0 for batchnorm and bias tensors is recommended
- since there is no API to freeze part of the model yet, using a learning rate of 0 for some tensors provided this functionality.

This was done in the stateful optimizer in SwiftAI notebook 09 with the old keyPath approach but doesn't seem possible anymore (please correct me if I'm wrong).

I'm all for not using keyPaths but we then need a way to define a `learningRate` object that would have different values for each trainable parameter of the model (and same for other hyper-parameters).

**Comment by eaplatanios** [comment]
I believe this is not due to the removal of `AllDifferentiableVariables` but rather the switch to non-key-path-based optimizers. You will probably have to make your learning rate have type `TangentVector` so that it contains a separate value for each model parameter.

**Comment by rxwei** [comment]
@eaplatanios is right. You can try making your learning rates have type TangentVector. I’m happy to help look at the notebooks on Monday as well.

**Comment by sgugger** [comment]
While there may be a workaround for a thing like having a different learning rate for each parameter, I believe there is a bigger problem.

First the design is working really well for simple examples, but you will always hit walls when a new optimizer comes along and requires a more complicated function of the gradients, so you'll always need the ability to index in the tensors. For instance, [AdaMax](https://github.com/tensorflow/swift-apis/blob/0193aaf62396e254c9af55740f91d4efd0f847c2/Sources/TensorFlow/Optimizers/MomentumBased.swift#L230) is currently implemented this way while waiting for max to be available for `TangentVector`.
In [LARS](https://arxiv.org/abs/1708.03888) or [LAMB](https://arxiv.org/abs/1904.00962) you will need to take the L2 norm for each tensor, which will require another new addition of functionality to `TangentVector`, then whatever new optimizer comes next will probably require some new one as well... This is just to say that the `keyPath` approach goes away for simple things, but we need it as a backup plan (or another way to index in the model/gradients) for more complicated stuff. 

But thinking about LARS and LAMB made me realize there is a very big flaw in the current design: it assumes the update will be computed using the gradients only. Those optimizers use a combination of the gradients *and* the weights. As does something as simple as true weight decay (see [AdamW](https://arxiv.org/abs/1711.05101)). So either you need to be able to:
- mix `model` and `direction` in the formula giving the update of the optimizer, for instance:
```swift
model.move(along: -learningRate * (direction + weightDecay * model))
```
- or index conjointly into the `model` tensors and the `gradient` tensors (which is what `allDifferentiableVariables` provided)

In the first case, we need `Model.TangentVector == Model`, and since `allDifferentiableVariables` is deprecated and points to `Self`, we need that in the second case too. Problem is, it's not true, even for the simplest models.

**Comment by eaplatanios** [comment]
@sgugger I totally agree about the strong limitations of the current design. I previously opened #325 to track part of this problem and I recently also started [this discussion](https://groups.google.com/a/tensorflow.org/forum/#!topic/swift/DKqBXWwMz7I) that focuses on more use cases of `KeyPathIterable` with hopes of arriving at a better design. Given my Scala background, the most elegant solution I can think of is using type classes (which can be represented using protocols) and allowing for automatic derivations of type classes by letting the user define derivation rules. In fact, that's what I was doing in [TF Scala](https://github.com/eaplatanios/tensorflow_scala) for a lot of use cases. However, Swift does not currently support "customized" automatic derivation of protocol conformances (some derivations are hardcoded in the compiler). I would love to see either language-level support for protocol conformances derivation or some other way to better handle these use cases. The issue or the Google Group discussion I mentioned may be a good place to continue this discussion and increase the community's interest in this issue.

**Comment by rxwei** [comment]
After some offline discussions with @sgugger, @mattjj and @dougalm, I think we can use a refined protocol to resume the `AllDifferentiableVariables` behavior with significantly simpler semantics that what we had before.

```swift
/// A type representing the product manifold of a vector space with something else,
/// where `TangentVector` is equal to the vector space part of the product manifold.
public protocol PartialVectorSpaceDifferentiable: Differentiable {
    var vectorView: TangentVector { get }
}
```

With this, you can write
```swift
model.move(along: -learningRate * (direction + weightDecay * model.vectorView))
```

The name `PartialVectorSpaceDifferentiable` is not great. Any suggestions are of course welcome.

**Comment by eaplatanios** [comment]
@rxwei Thanks for the update! Why not make `vectorView` always part of `Differentiable`?

**Comment by rxwei** [comment]
`vectorView` is defined as "the `V` part of product manifold `V x M` where `(V x M).Tan == V`". This does not make sense for general manifolds (e.g. a sphere).

**Comment by sgugger** [comment]
Wrt to naming, it sounds to me like `vectorView` makes sense when we are trying to differentiate in an euclidean space (maybe a larger class of spaces though). So maybe `EuclideanDifferentiable`?

Another possibility is `DifferentiableWithProjector` since `tangentVector` is a projector of `Self` to `TangentVector`

**Comment by rxwei** [comment]
`EuclideanDifferentiable` actually sounds good.
`DifferentiableWithProjector` is weird in that there's a mathematical notion of "tangent space projection" (search for "tangent space projector" in [the manopt tutorial](https://www.manopt.org/tutorial.html)) and it is completely different from what `vectorView` does.

**Comment by rxwei** [comment]
#26827 adds the protocol, which you can already conform to manually in fast.ai notebooks. I'll implement derived conformances for `vectorView` on Monday.

**Comment by eaplatanios** [comment]
That’s great. Thanks Richard!

**Comment by sgugger** [comment]
Thanks Richard!
Re- conforming a layer manually to `EuclidianDifferentiable`. Not sure what the best way is to define a TangentVector. For a `Dense` layer (where the differentiable part is weight and bias), this:
``` swift
public var vectorView: TangentVector {
        _read   { 
            var res = TangentVector.zero
            res.weight = weight
            res.bias = bias
            yield res
        }
        _modify { 
            var res = TangentVector.zero
            res.weight = weight
            res.bias = bias
            yield &res
        }
    }
```
works but this
``` swift
public var vectorView: TangentVector {
        _read   { yield TangentVector(weight: weight, bias: bias) }
        _modify { yield &TangentVector(weight: weight, bias: bias) }
}
```
crashes with a type checker error. Is there a simpler way to create the proper `vectorView`? Ultimately it would be great if it just picked the differentiable attributes automatically.
